# Import Data and Libraries

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Set path
path = r'C:\Users\asus\Instacart Basket Analysis 01-2024'

In [3]:
# import dataframe from previous Task

df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# STEP 4 - Wrangle Data

## renaming columns

In [5]:
# renaming columns

df.rename(columns = {'Surnam' : 'Surname', 'STATE' : 'State', 'n_dependants': 'Dependants', 'date_joined' : 'Date_joined', 'fam_status' : 'Family_status'}, inplace = True)

In [6]:
df.head()

user_id First Name   Surname  Gender       State  Age Date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Dependants Family_status  income  
0           3       married  165665  
1           0        single   59285  
2           2       married   99568  
3           0        single   42049  
4           1       married   40374

# STEP 5 - Data Quality and Consistency Checks

## checking for missing values

In [7]:
# checking for missing values

df.value_counts(dropna=False)

user_id  First Name  Surname   Gender  State          Age  Date_joined  Dependants  Family_status                     income
1        Linda       Nguyen    Female  Alabama        31   2/17/2019    3           married                           40423     1
137478   Elizabeth   Mcdonald  Female  Maryland       19   12/16/2019   2           living with parents and siblings  28625     1
137468   Todd        Welch     Male    Nevada         31   7/11/2017    1           married                           42533     1
137469   Douglas     Andrews   Male    New Hampshire  45   4/18/2018    1           married                           125718    1
137470   Jeremy      Bennett   Male    New Jersey     19   5/10/2019    3           living with parents and siblings  79969     1
                                                                                                                               ..
68740    Michael     Wall      Male    Tennessee      48   8/18/2017    3           married    

In [8]:
df['State'].value_counts(dropna=False)

State
Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana 

In [9]:
df['First Name'].value_counts(dropna=False)

First Name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [10]:
df['Surname'].value_counts(dropna=False)

Surname
Hamilton      252
Randall       248
Lamb          243
Pennington    243
Barnett       242
             ... 
Poole         172
Bauer         166
Pearson       164
Payne         163
Jordan        162
Name: count, Length: 1000, dtype: int64

In [11]:
df['Gender'].value_counts(dropna=False)

Gender
Male      104067
Female    102142
Name: count, dtype: int64

In [12]:
df['Age'].value_counts(dropna=False)

Age
19    3329
55    3317
51    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: count, Length: 64, dtype: int64

In [13]:
df['Date_joined'].value_counts(dropna=False)

Date_joined
9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: count, Length: 1187, dtype: int64

In [14]:
df['Dependants'].value_counts(dropna=False)

Dependants
0    51602
3    51594
1    51531
2    51482
Name: count, dtype: int64

In [15]:
df['Family_status'].value_counts(dropna=False)

Family_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [16]:
df['income'].value_counts(dropna=False)

income
57192     10
95891     10
95710     10
97532      9
98675      9
          ..
73141      1
71524      1
74408      1
44780      1
148828     1
Name: count, Length: 108012, dtype: int64

In [17]:
df.isnull().sum()

user_id              0
First Name       11259
Surname              0
Gender               0
State                0
Age                  0
Date_joined          0
Dependants           0
Family_status        0
income               0
dtype: int64

there are 11,259 missing first names in the dataframe, but we will keep them as is since last names still match and deleting this data could skew negatively reults of any future data searches.

## Checking for duplicates

In [18]:
# checking for duplicates

df_dups = df[df.duplicated()]

In [19]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Surname, Gender, State, Age, Date_joined, Dependants, Family_status, income]
Index: []

no duplicates found

## Checking for mixed types data

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        206209 non-null  int64 
 1   First Name     194950 non-null  object
 2   Surname        206209 non-null  object
 3   Gender         206209 non-null  object
 4   State          206209 non-null  object
 5   Age            206209 non-null  int64 
 6   Date_joined    206209 non-null  object
 7   Dependants     206209 non-null  int64 
 8   Family_status  206209 non-null  object
 9   income         206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [21]:
# checking mixed types of data

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

First Name


C:\Users\Asus\AppData\Local\Temp\ipykernel_2576\791970356.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Asus\AppData\Local\Temp\ipykernel_2576\791970356.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Asus\AppData\Local\Temp\ipykernel_2576\791970356.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Asus\AppData\Local\Temp\ipykernel_2576\791970356.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Asus\AppData\Local\Temp\ipykernel_2576\791970356.py:4: FutureWarning: D

no mixed type data found

In [22]:
df.shape

(206209, 10)

In [23]:
df.describe()

user_id            Age     Dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [24]:
# changing the Date Joined data type from 'Object' to 'Datetime'

df['Date_joined']=pd.to_datetime(df['Date_joined'])

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   user_id        206209 non-null  int64         
 1   First Name     194950 non-null  object        
 2   Surname        206209 non-null  object        
 3   Gender         206209 non-null  object        
 4   State          206209 non-null  object        
 5   Age            206209 non-null  int64         
 6   Date_joined    206209 non-null  datetime64[ns]
 7   Dependants     206209 non-null  int64         
 8   Family_status  206209 non-null  object        
 9   income         206209 non-null  int64         
dtypes: datetime64[ns](1), int64(4), object(5)
memory usage: 15.7+ MB


In [26]:
df.head(1000)

user_id First Name   Surname  Gender           State  Age Date_joined  \
0      26711    Deborah  Esquivel  Female        Missouri   48  2017-01-01   
1      33890   Patricia      Hart  Female      New Mexico   36  2017-01-01   
2      65803    Kenneth    Farley    Male           Idaho   35  2017-01-01   
3     125935   Michelle     Hicks  Female            Iowa   40  2017-01-01   
4     130797        Ann   Gilmore  Female        Maryland   26  2017-01-01   
..       ...        ...       ...     ...             ...  ...         ...   
995   157936   Nicholas     Gross    Male    Rhode Island   52  2017-01-06   
996    21309     Ernest     Arias    Male    South Dakota   64  2017-01-06   
997    82291      Frank  Callahan    Male  North Carolina   49  2017-01-06   
998    77675      Frank     Lloyd    Male          Alaska   71  2017-01-06   
999   200662   Margaret      Hale  Female  North Carolina   72  2017-01-06   

     Dependants     Family_status  income  
0             3           married  165665  
1             0            single   59285  
2             2           married   99568  
3             0            single   42049  
4             1           married   40374  
..          ...               ...     ...  
995           2           married  147933  
996           1           married  143108  
997           1           married  135887  
998           1           married   92629  
999           0  divorced/widowed   88404  

[1000 rows x 10 columns]

# STEP 6 - combine current Customer Data with Current Instacard Data

In [27]:
# import instacard dataframe as df_two

df_two = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_new2.pkl'))

In [28]:
df_two.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int64   
 1   product_name            object  
 2   aisle_id                int64   
 3   department_id           int64   
 4   prices                  float64 
 5   order_id                int32   
 6   user_id                 int32   
 7   order_number            int8    
 8   orders_day_of_week      int8    
 9   order_time_of_day       int8    
 10  days_since_prior_order  float16 
 11  add_to_cart_order       int32   
 12  reordered               int8    
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  Busiest_days            object  
 17  busiest_period_of_day   object  
 18  max_order               int8    
 19  loyalty_flag            object  
 20  spender_average         float64 
 21  spende

In [29]:
# Musst change user_id of df from int64 to int32 to match dataframes

df['user_id'] = df['user_id'].astype('int32')

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   user_id        206209 non-null  int32         
 1   First Name     194950 non-null  object        
 2   Surname        206209 non-null  object        
 3   Gender         206209 non-null  object        
 4   State          206209 non-null  object        
 5   Age            206209 non-null  int64         
 6   Date_joined    206209 non-null  datetime64[ns]
 7   Dependants     206209 non-null  int64         
 8   Family_status  206209 non-null  object        
 9   income         206209 non-null  int64         
dtypes: datetime64[ns](1), int32(1), int64(3), object(5)
memory usage: 14.9+ MB


In [31]:
# drop the _merge column from df_two as it is not needed and could cause issues

df_two = df_two.drop(columns = ['_merge'])

In [32]:
df_two.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   product_id              int64  
 1   product_name            object 
 2   aisle_id                int64  
 3   department_id           int64  
 4   prices                  float64
 5   order_id                int32  
 6   user_id                 int32  
 7   order_number            int8   
 8   orders_day_of_week      int8   
 9   order_time_of_day       int8   
 10  days_since_prior_order  float16
 11  add_to_cart_order       int32  
 12  reordered               int8   
 13  price_range_loc         object 
 14  busiest_day             object 
 15  Busiest_days            object 
 16  busiest_period_of_day   object 
 17  max_order               int8   
 18  loyalty_flag            object 
 19  spender_average         float64
 20  spender_flag            object 
 21  frequency_average       float

In [33]:
#check output of df

df.shape

(206209, 10)

In [34]:
# check output of df_two

df_two.shape

(32404859, 23)

In [35]:
# merge both dataframes df and df_two using "user_id" column

df_merged_large = df.merge(df_two, on = 'user_id', indicator = True)

In [36]:
df.shape

(206209, 10)

In [37]:
df_two.shape

(32404859, 23)

In [38]:
df_merged_large.shape

(32404859, 33)

In [40]:
df_merged_large.head(5)

user_id First Name   Surname  Gender     State  Age Date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
1    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
2    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
3    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
4    26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   

   Dependants Family_status  income  ...     busiest_day    Busiest_days  \
0           3       married  165665  ...  Regularly busy    Busiest days   
1           3       married  165665  ...  Regularly busy  Regularly busy   
2           3       married  165665  ...  Regularly busy    Busiest days   
3           3       married  165665  ...  Regularly busy  Regularly busy   
4           3       married  165665  ...  Regularly busy    Slowest days   

   busiest_period_of_day  max_order  loyalty_flag  spender_average  \
0            Most Orders          8  New customer         7.988889   
1            Most Orders          8  New customer         7.988889   
2            Most Orders          8  New customer         7.988889   
3            Most Orders          8  New customer         7.988889   
4            Most Orders          8  New customer         7.988889   

   spender_flag  frequency_average    frequency_flag  _merge  
0   low spender               19.0  Regular customer    both  
1   low spender               19.0  Regular customer    both  
2   low spender               19.0  Regular customer    both  
3   low spender               19.0  Regular customer    both  
4   low spender               19.0  Regular customer    both  

[5 rows x 33 columns]

In [41]:
df_merged_large.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_cust_merge.pkl'))